base/compiler/typeinfer.jl

```diff
@@ -86,12 +86,17 @@ Empty out the previously recorded type inference timings (`Core.Compiler._timing
 start the ROOT() timer again. `ROOT()` measures all time spent _outside_ inference.
 """
 function reset_timings()
-    empty!(_timings)
-    push!(_timings, Timing(
-        # The MethodInstance for ROOT(), and default empty values for other fields.
-        InferenceFrameInfo(ROOTmi, 0x0, Any[], Any[Core.Const(ROOT)], 1),
-        _time_ns()))
-    return nothing
+    ccall(:jl_typeinf_lock_begin, Cvoid, ())
+    try
+        empty!(_timings)
+        push!(_timings, Timing(
+            # The MethodInstance for ROOT(), and default empty values for other fields.
+            InferenceFrameInfo(ROOTmi, 0x0, Any[], Any[Core.Const(ROOT)], 1),
+            _time_ns()))
+        return nothing
+    finally
+        ccall(:jl_typeinf_lock_end, Cvoid, ())
+    end
```

Comment:
Oh dip, it turns out that we've already done some work to shrink the inference lock (!! 🎉 exciting!! 🎉), so i can't rely on it for locking the inference profiling anymore. 😮 😢 
https://github.com/JuliaLang/julia/pull/46324

Gonna move this PR to a draft. Needs more thinking now. Probably we at least need a data lock around the `_timings` buffer, but also the operation of the inference profiling was heavily assuming that inference was non-concurrent, which it no longer is.